src/dlink.h

```diff
@@ -25,10 +25,13 @@ class dlink_list
 public:
     dlink_node *head = nullptr;
     dlink_node *tail = nullptr;
+    void add(void *data, dlink_node *node);
+    void addAfter(void *data, dlink_node *node, dlink_node *reference);
+    void addTail(void *data, dlink_node *node);
+    void moveToFront(dlink_node *node);
+    void remove(dlink_node *node);
```

Comment:
>  Add class methods to dlink_list

Please remove these changes from this PR.

IMO, adding whole-list-manipulation methods to an individual list element is a step in the wrong direction because it obscures the different roles those two concepts (i.e. the whole list and an individual element) play. It makes wrong-role mistakes (e.g., supplying an inner node instead of the list head) easier to make and more difficult to spot during review.

We can (and probably should) convert the dlink_list (i.e., invasive list with nodes that may not belong to any list) API to modern C++. However, doing so requires a different approach and should be done in a stand-alone PR. An ideal PR would also minimize existing caller changes if such minimization is possible, but I have not research that possibility. If you decide to write such a dedicated PR, I recommend posting the proposed class declarations first along with a couple of representative user code adjustments, before syncing the rest of Squid code.

P.S. I did not review individual method implementations.

Reply:
Maybe there is a misunderstanding. These class methods are essentially aliases for the free ones; they are so far only syntactic sugar. I noticed that it is a common pattern for dlink_list users is to have the dlink_node as a data member and to use it e.g. for O(1) LRU.
In code:
```
struct foo {
  dlink_node dl;
}
// ..
void foo::some_method() {
  // free function
  dlink_add(this, &dl, &some_list);
  // member function, same effect
  some_list.add(this, &dl);
}
```

Reply:
Please interpret my change request as if I knew that "These class methods are essentially aliases for the free ones". I did not check, but that was indeed my assumption for all the methods but moveToFront(). That single exception is irrelevant here.

There are often many ways to achieve the same assembly code. In most non-trivial cases, very few of those ways are actually the right thing to do at the C++ level. The proposed way is not the right thing (for the reasons stated earlier), and the invasive list API problem itself deserves a dedicated PR (even if you disagree with those earlier reasons).

Reply:
Ok, removing and converting moveToFront to free function